mmcls/core/hook/NumClassCheckHook.py

```diff
@@ -0,0 +1,55 @@
+# Copyright (c) OpenMMLab. All rights reserved
+from mmcv.runner.hooks import Hook
+
+from .builder import HOOKS
+
+
+@HOOKS.register_module()
+class NumClassCheckHook(Hook):
+
+    def _check_head(self, runner):
+        """Check whether the `num_classes` in head matches the length of
+        `CLASSES` in `dataset`.
+
+        Args:
+            runner (obj:`EpochBasedRunner`): Epoch based Runner.
+        """
+        model = runner.model
+        dataset = runner.data_loader.dataset
+        if dataset.CLASSES is None:
+            runner.logger.warning(
+                f'Please set `CLASSES` '
+                f'in the {dataset.__class__.__name__} and'
+                f'check if it is consistent with the `num_classes` '
+                f'of head')
+        else:
+            assert type(dataset.CLASSES) is not str, \
+                (f'`CLASSES` in {dataset.__class__.__name__}'
+                 f'should be a tuple of str.'
+                 f'Add comma if number of classes is 1 as '
+                 f'CLASSES = ({dataset.CLASSES},)')
+            for name, module in model.named_modules():
+                if hasattr(module, 'num_classes'):
+                    assert module.num_classes == len(dataset.CLASSES), \
+                        (f'The `num_classes` ({module.num_classes}) in '
+                         f'{module.__class__.__name__} of '
+                         f'{model.__class__.__name__} does not matches '
+                         f'the length of `CLASSES` '
+                         f'{len(dataset.CLASSES)}) in '
+                         f'{dataset.__class__.__name__}')
+
+    def before_train_epoch(self, runner):
+        """Check whether the training dataset is compatible with head.
+
+        Args:
+            runner (obj:`EpochBasedRunner`): Epoch based Runner.
+        """
+        self._check_head(runner)
+
+    def before_val_epoch(self, runner):
+        """Check whether the dataset in val epoch is compatible with head.
+
+        Args:
+            runner (obj:`EpochBasedRunner`): Epoch based Runner.
+        """
+        self._check_head(runner)
```

Comment:
I think simply checking once in "before_run" is enough. Is it necessary to check it before each epoch?
And it won't work if users use `IterBasedRunner`

Reply:
I just noticed that `before_run` hook can't be used in this case, because the `runner.data_loader` variable is populated when the runner has already entered the train/eval stage, but the `before_run` hook is called before this has happened.

Reply:
I just noticed that `before_run` hook can't be used in this case, because the `runner.data_loader` variable is populated when the runner has already entered the train/eval stage, but the `before_run` hook is called before this has happened.

Reply:
Got it.
The problem is the dataloader is not assigned to the `runner.dataloader` before `before_run` hook is called. It's also an mmcv design problem.